File: haskell/anagram/src/Anagram.hs
module Anagram (anagramsFor) where

import Data.List (sort)
import Data.Char (toLower)

anagramsFor :: String -> [String] -> [String]
anagramsFor a = filter (anagram a)

anagram :: [Char] -> [Char] -> Bool
anagram s1 s2 = lc1 /= lc2 && sort lc1 == sort lc2
    where
        lc1 = map toLower s1
        lc2 = map toLower s2

File: cpp/anagram/anagram.cpp
#include "anagram.h"

#include "anagram.h"
#include <algorithm>
#include <cctype>

namespace anagram {
    anagram::anagram(const string &word) {
        lc = word;
        tolower(lc);
        letters = lc;
        sort(letters.begin(), letters.end());
    }

    void anagram::tolower(string &basicString) { transform(basicString.begin(), basicString.end(), basicString.begin(), ::tolower); }

    vector <string> anagram::matches(const vector <string> &cands) {
        vector <string> r;
        for (const string &s: cands) {
            string dup = s;
            tolower(dup);
            if (dup == lc) continue;
            sort(dup.begin(), dup.end());
            if (dup == letters) r.emplace_back(s);
        }
        return r;
    }
}
